packages/typespec-client-generator-core/test/decorators.test.ts

```diff
@@ -4256,4 +4256,265 @@ describe("typespec-client-generator-core: decorators", () => {
+      const sdkPackage = runner.context.sdkPackage;
+      const client = sdkPackage.clients[0];
+      strictEqual(client.methods.length, 1);
+      const method = client.methods[0];
+      strictEqual(method.kind, "basic");
+      strictEqual(method.name, "func");
+      strictEqual(method.parameters.length, 3);
+
+      const contentTypeParam = method.parameters.find((x) => x.name === "contentType");
+      ok(contentTypeParam);
+
+      const fooParam = method.parameters.find((x) => x.name === "foo");
+      ok(fooParam);
+
+      const barParam = method.parameters.find((x) => x.name === "bar");
+      ok(barParam);
+
+      const httpOp = method.operation;
+      strictEqual(httpOp.parameters.length, 1);
+      strictEqual(httpOp.parameters[0].correspondingMethodParams[0], contentTypeParam);
+
+      ok(httpOp.bodyParam);
+      strictEqual(httpOp.bodyParam.correspondingMethodParams.length, 2);
+      strictEqual(httpOp.bodyParam.correspondingMethodParams[0], fooParam);
+      strictEqual(httpOp.bodyParam.correspondingMethodParams[1], barParam);
+
+      const runnerWithCsharp = await createSdkTestRunner({
+        emitterName: "@azure-tools/typespec-csharp",
+      });
+      await runnerWithCsharp.compileWithCustomization(mainCode, customizationCode);
+      ok(runnerWithCsharp.context.sdkPackage.models.find((x) => x.name === "Params"));
+
+      const sdkPackageWithCsharp = runnerWithCsharp.context.sdkPackage;
+      strictEqual(sdkPackageWithCsharp.clients.length, 1);
+
+      strictEqual(sdkPackageWithCsharp.clients[0].methods.length, 1);
+      const methodWithCsharp = sdkPackageWithCsharp.clients[0].methods[0];
+      strictEqual(methodWithCsharp.kind, "basic");
+      strictEqual(methodWithCsharp.name, "func");
+      strictEqual(methodWithCsharp.parameters.length, 2);
+      const contentTypeParamWithCsharp = methodWithCsharp.parameters.find(
+        (x) => x.name === "contentType"
+      );
+      ok(contentTypeParamWithCsharp);
+
+      const paramsParamWithCsharp = methodWithCsharp.parameters.find((x) => x.name === "params");
+      ok(paramsParamWithCsharp);
+      strictEqual(
+        sdkPackageWithCsharp.models.find((x) => x.name === "Params"),
+        paramsParamWithCsharp.type
+      );
+
+      const httpOpWithCsharp = methodWithCsharp.operation;
+      strictEqual(httpOpWithCsharp.parameters.length, 1);
+      strictEqual(
+        httpOpWithCsharp.parameters[0].correspondingMethodParams[0],
+        contentTypeParamWithCsharp
+      );
+      ok(httpOpWithCsharp.bodyParam);
+      strictEqual(httpOpWithCsharp.bodyParam.correspondingMethodParams.length, 1);
+      strictEqual(httpOpWithCsharp.bodyParam.correspondingMethodParams[0], paramsParamWithCsharp);
+    });
+
+    it("regrouping", async () => {
+      const mainCode = `
+        @service
+        namespace MyService;
+        model Params {
+          foo: string;
+          bar: string;
+          fooBar: string;
+        }
```

Comment:
Can we mix a `@query` or `@header` into this model? It is pretty common to have a POST that contains body and a few query param or header param.

One uncertainty is whether the logic of `compareModelProperties` can handle some query/header that was not explicitly defined in service typespec.
For example, service API may have conditional headers like `ifMatch`. And these would be from typespec-azure-core [trait](https://github.com/Azure/typespec-azure/blob/main/packages/typespec-azure-core/lib/traits.tsp#L163-L176), where [headers](https://github.com/Azure/typespec-azure/blob/main/packages/typespec-azure-core/lib/models.tsp#L136-L157)
<-- it is fine we handle this later. We can get the PR merged for simple "revert the spread" scenario.

Reply:
a related concern is that is the code could support override templated operation like `op scheduleRepairs is Operations.ResourceAction<Widget, RepairInfo, RepairResult>;`